metaflow/plugins/argo/argo_inner_flow.py

```diff
@@ -34,7 +67,104 @@ def __init__(
         self._argo_run_id = self._flow_information["metadata"]["name"]
         self._metaflow_run_id = f"argo-{self._argo_run_id}"
         self._kubernetes_namespace = self._flow_information["metadata"]["namespace"]
-        self._exception = None
-        self._status = None
 
-        print(f"Inner flow has started w/ Argo id: {self._argo_run_id}")
+        # Waiting for Argo workflow to trigger is not optional.
+        # It is necessary to determine Flow name if not given as parameter
+        wait_to_trigger = 20  # wait time is 20 seconds
+        print("attempting to trigger Argo workflow")
+        while wait_to_trigger > 0 and self.status is None:
+            time.sleep(1)
+            wait_to_trigger -= 1
+
+        if wait_to_trigger == 0:
+            raise Exception("Inner flow failed to begin running")
+
+        print(
+            f"A run of Metaflow flow {self._flow_name} has started w/ Argo id: {self._argo_run_id}"
+        )
+
+        # [optional] wait for argo workflow to finish running
+        if self._wait:
+            print("\nNow we will wait for the flow to finish")
+            wait_remaining = self._wait_to_run  # in minutes (5s for testing)
+            print(f"status: {self.status}, 'mins' to wait: {wait_remaining}")
+            while wait_remaining > 0 and self.status == "Running":
+                wait_remaining -= 1
+                print(f"status: {self.status}, 'mins' to wait: {wait_remaining}")
+                time.sleep(5)
+
+            if wait_remaining == 0:
+                print("Inner flow timed out. better work on that speed for next time!")
+
+            print("\nInner flow is finished!!!")
+
+        else:
+            print("Not waiting for inner flow to finish")
+
+            
+    @property
+    def status(self):
+        wf = self._argo_client.get_workflow(self._argo_run_id)
+
+        if "status" in wf and "phase" in wf["status"]:
+            self._status = wf["status"]["phase"]
+
+        return self._status
+
+    
+    @property
+    def failed_steps(self):
+        if self.status != "Failed":
+            return []
+
+        failed_steps = []
+
+        wf = self._argo_client.get_workflow(self._argo_run_id)
+        nodes_info = wf["status"]["nodes"]
+        for node in nodes_info:
+            if (
+                nodes_info[node]["phase"] == "Failed"
+                and nodes_info[node]["type"] == "Pod"
+            ):
+                failed_steps.append(nodes_info[node]["templateName"])
+
+        self._failed_steps = failed_steps
+
+        return self._failed_steps
+
+    
+    def _find_metaflow_run(self, metaflow_run_location):
+        from metaflow import Run, namespace
+        
+        namespace(None)
+        metaflow_run_location = self._flow_name + "/" + self._metaflow_run_id
+        attempts_to_find_metaflow_run = 60  # gives 1 minute to find metaflow run
```

Comment:
60 seconds feel long to me. If we know a flow fails, metaflow wouldn't get more info in this 60 second; if the flow is still running there shouldn't be an exception.

Reply:
Thanks, that makes sense. I added a check to see if still running, and I return None if it's still running. Otherwise, it checks for 5 seconds now